eosmetrics/emtr-event-recorder.c

```diff
@@ -1277,4 +1278,48 @@ emtr_event_recorder_record_stop_sync (EmtrEventRecorder *self,
                TRUE /* is_synchronous */);
 }
 
+/**
+ * emtr_event_recorder_start_aggregate_timer:
+ * @self: an #EmtrEventRecorder
+ * @event_id: an RFC 4122 UUID representing the type of event that took place
+ * @aggregate_key: the identifier used to aggregate data
+ * @auxiliary_payload: miscellaneous data to associate with the events. Must
+ * not contain maybe variants as they are not compatible with D-Bus.
+ *
+ * Requests the metrics daemon to create an aggregate timer.
+ *
+ * Returns: (transfer full)(nullable): a #EmtrAggregateTimer
+ */
+EmtrAggregateTimer *
+emtr_event_recorder_start_aggregate_timer (EmtrEventRecorder *self,
+                                           const gchar       *event_id,
+                                           GVariant          *aggregate_key,
+                                           GVariant          *auxiliary_payload)
+{
+  EmtrEventRecorderPrivate *priv =
+    emtr_event_recorder_get_instance_private (self);
+  GVariantBuilder event_id_builder;
+  uuid_t parsed_event_id;
+
+  g_return_val_if_fail (EMTR_IS_EVENT_RECORDER (self), NULL);
+  g_return_val_if_fail (_IS_VARIANT (aggregate_key), NULL);
+  g_return_val_if_fail (_IS_VARIANT (auxiliary_payload), NULL);
```

Comment:
I don't think these should need to contain variants.

Here is the current code in our gnome-shell fork:

```
src/shell-app-system.c:632:          emtr_event_recorder_record_start (emtr_event_recorder_get_default (),
src/shell-app-system.c-633-                                            SHELL_APP_IS_OPEN_EVENT,
src/shell-app-system.c-634-                                            g_variant_new_string (app_address),
src/shell-app-system.c-635-                                            g_variant_new_string (app_info_id));
```

So I would expect the following call to succeed:

```c
  timer = emtr_event_recorder_start_aggregate_timer (fixture->recorder,
                                                     MEANINGLESS_EVENT,
                                                     g_variant_new_string ("org.gnome.Builder.desktop"),
                                                     g_variant_new_string ("org.gnome.Builder.desktop"));
```

But in fact it fails.

When calling the D-Bus method, then yes, the D-Bus argument type is `v` so the app-provided variant needs to be boxed with `g_variant_new_variant`.

Reply:
In the D-Bus API, the payload is optional (faking the maybe type with a boolean plus a normal variant which is set to a dummy value). In the existing C API for existing methods, this is supported by allowing a NULL payload. For example:

```
static void
test_event_recorder_record_event_sync (struct RecorderFixture *fixture,
                                       gconstpointer           unused)
{
  emtr_event_recorder_record_event_sync (fixture->recorder, MEANINGLESS_EVENT,
    NULL);
}
```

Again I think we should preserve this property for application convenience.